File: compiler/src/checker/const_eval.c
#include "const_eval.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <limits.h>

// 检查表达式是否为常量
int const_is_constant(ASTNode *expr) {
    if (!expr) return 0;
    
    switch (expr->type) {
        case AST_NUMBER:
        case AST_BOOL:
        case AST_STRING:
            return 1;
        case AST_IDENTIFIER:
            // 变量引用不是常量（除非是const变量，但需要符号表信息）
            return 0;
        case AST_BINARY_EXPR: {
            // 递归检查左右操作数
            int left_const = const_is_constant(expr->data.binary_expr.left);
            int right_const = const_is_constant(expr->data.binary_expr.right);
            return left_const && right_const;
        }
        case AST_UNARY_EXPR: {
            // 递归检查操作数
            return const_is_constant(expr->data.unary_expr.operand);
        }
        default:
            return 0;
    }
}

// 获取常量整数值
long long const_get_int_value(ASTNode *expr) {
    if (!expr) return 0;
    
    if (expr->type == AST_NUMBER) {
        char *endptr;
        long long val = strtoll(expr->data.number.value, &endptr, 10);
        if (*endptr == '\0' && errno != ERANGE) {
            return val;
        }
    } else if (expr->type == AST_BOOL) {
        return expr->data.bool_literal.value ? 1 : 0;
    }
    
    return 0;
}

// 常量表达式求值
int const_eval_expr(ASTNode *expr, ConstValue *result) {
    if (!expr || !result) return 0;
    
    result->type = CONST_VAL_NONE;
    
    switch (expr->type) {
        case AST_NUMBER: {
            char *endptr;
            errno = 0;
            long long int_val = strtoll(expr->data.number.value, &endptr, 10);
            
            // 检查是否是整数
            if (*endptr == '\0' && errno != ERANGE) {
                result->type = CONST_VAL_INT;
                result->value.int_val = int_val;
                return 1;
            }
            
            // 尝试作为浮点数解析
            errno = 0;
            double float_val = strtod(expr->data.number.value, &endptr);
            if (*endptr == '\0' && errno != ERANGE) {
                result->type = CONST_VAL_FLOAT;
                result->value.float_val = float_val;
                return 1;
            }
            
            return 0;
        }
        
        case AST_BOOL: {
            result->type = CONST_VAL_BOOL;
            result->value.bool_val = expr->data.bool_literal.value;
            return 1;
        }
        
        case AST_BINARY_EXPR: {
            ConstValue left_val, right_val;
            int left_const = const_eval_expr(expr->data.binary_expr.left, &left_val);
            int right_const = const_eval_expr(expr->data.binary_expr.right, &right_val);
            
            if (!left_const || !right_const) {
                return 0;  // 不是常量表达式
            }
            
            // 根据操作符类型进行运算
            int op = expr->data.binary_expr.op;
            
            // 整数运算
            if (left_val.type == CONST_VAL_INT && right_val.type == CONST_VAL_INT) {
                long long a = left_val.value.int_val;
                long long b = right_val.value.int_val;
                long long result_val;
                int overflow = 0;
                
                switch (op) {
                    case 36: // TOKEN_PLUS
                        result_val = a + b;
                        // 检查溢出（简化版）
                        if ((a > 0 && b > 0 && result_val < a) ||
                            (a < 0 && b < 0 && result_val > a)) {
                            overflow = 1;
                        }
                        break;
                    case 37: // TOKEN_MINUS
                        result_val = a - b;
                        if ((a > 0 && b < 0 && result_val < a) ||
                            (a < 0 && b > 0 && result_val > a)) {
                            overflow = 1;
                        }
                        break;
                    case 38: // TOKEN_ASTERISK
                        result_val = a * b;
                        // 简化的溢出检查
                        if (a != 0 && result_val / a != b) {
                            overflow = 1;
                        }
                        break;
                    case 39: // TOKEN_SLASH
                        if (b == 0) {
                            return 0;  // 除零错误
                        }
                        result_val = a / b;
                        break;
                    case 40: // TOKEN_PERCENT
                        if (b == 0) {
                            return 0;  // 除零错误
                        }
                        result_val = a % b;
                        break;
                    case 57: // TOKEN_EQUAL
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a == b);
                        return 1;
                    case 58: // TOKEN_NOT_EQUAL
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a != b);
                        return 1;
                    case 59: // TOKEN_LESS
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a < b);
                        return 1;
                    case 60: // TOKEN_LESS_EQUAL
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a <= b);
                        return 1;
                    case 61: // TOKEN_GREATER
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a > b);
                        return 1;
                    case 62: // TOKEN_GREATER_EQUAL
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a >= b);
                        return 1;
                    default:
                        return 0;  // 不支持的操作符
                }
                
                if (overflow) {
                    return 0;  // 溢出，不是有效常量
                }
                
                result->type = CONST_VAL_INT;
                result->value.int_val = result_val;
                return 1;
            }
            
            // 浮点数运算
            if (left_val.type == CONST_VAL_FLOAT || right_val.type == CONST_VAL_FLOAT) {
                double a = (left_val.type == CONST_VAL_FLOAT) ? left_val.value.float_val : (double)left_val.value.int_val;
                double b = (right_val.type == CONST_VAL_FLOAT) ? right_val.value.float_val : (double)right_val.value.int_val;
                double result_val;
                
                switch (op) {
                    case 36: // TOKEN_PLUS
                        result_val = a + b;
                        break;
                    case 37: // TOKEN_MINUS
                        result_val = a - b;
                        break;
                    case 38: // TOKEN_ASTERISK
                        result_val = a * b;
                        break;
                    case 39: // TOKEN_SLASH
                        if (b == 0.0) {
                            return 0;  // 除零错误
                        }
                        result_val = a / b;
                        break;
                    case 57: // TOKEN_EQUAL
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a == b);
                        return 1;
                    case 58: // TOKEN_NOT_EQUAL
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a != b);
                        return 1;
                    case 59: // TOKEN_LESS
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a < b);
                        return 1;
                    case 60: // TOKEN_LESS_EQUAL
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a <= b);
                        return 1;
                    case 61: // TOKEN_GREATER
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a > b);
                        return 1;
                    case 62: // TOKEN_GREATER_EQUAL
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = (a >= b);
                        return 1;
                    default:
                        return 0;
                }
                
                result->type = CONST_VAL_FLOAT;
                result->value.float_val = result_val;
                return 1;
            }
            
            return 0;
        }
        
        case AST_UNARY_EXPR: {
            ConstValue operand_val;
            int operand_const = const_eval_expr(expr->data.unary_expr.operand, &operand_val);
            
            if (!operand_const) {
                return 0;
            }
            
            int op = expr->data.unary_expr.op;
            
            switch (op) {
                case 37: // TOKEN_MINUS (负号)
                    if (operand_val.type == CONST_VAL_INT) {
                        result->type = CONST_VAL_INT;
                        result->value.int_val = -operand_val.value.int_val;
                        return 1;
                    } else if (operand_val.type == CONST_VAL_FLOAT) {
                        result->type = CONST_VAL_FLOAT;
                        result->value.float_val = -operand_val.value.float_val;
                        return 1;
                    }
                    break;
                case 73: // TOKEN_EXCLAMATION (逻辑非)
                    if (operand_val.type == CONST_VAL_BOOL) {
                        result->type = CONST_VAL_BOOL;
                        result->value.bool_val = !operand_val.value.bool_val;
                        return 1;
                    }
                    break;
                default:
                    return 0;
            }
            
            return 0;
        }
        
        default:
            return 0;  // 不是常量表达式
    }
}

File: compiler-go/src/checker/const_eval.go
package checker

import (
	"strconv"

	"github.com/uya/compiler-go/src/lexer"
	"github.com/uya/compiler-go/src/parser"
)

// ConstValueType represents the type of a constant value
type ConstValueType int

const (
	ConstValInt ConstValueType = iota
	ConstValFloat
	ConstValBool
	ConstValNone // Not a constant
)

// ConstValue represents a constant value
type ConstValue struct {
	Type  ConstValueType
	Value struct {
		IntVal   int64
		FloatVal float64
		BoolVal  bool
	}
}

// EvalExpr evaluates a constant expression and stores the result
// Returns true if the expression is a constant, false otherwise
func EvalExpr(expr parser.Expr, result *ConstValue) bool {
	if expr == nil || result == nil {
		return false
	}

	result.Type = ConstValNone

	switch e := expr.(type) {
	case *parser.NumberLiteral:
		// Try to parse as integer first
		intVal, err := strconv.ParseInt(e.Value, 10, 64)
		if err == nil {
			result.Type = ConstValInt
			result.Value.IntVal = intVal
			return true
		}

		// Try to parse as float
		floatVal, err := strconv.ParseFloat(e.Value, 64)
		if err == nil {
			result.Type = ConstValFloat
			result.Value.FloatVal = floatVal
			return true
		}

		return false

	case *parser.BoolLiteral:
		result.Type = ConstValBool
		result.Value.BoolVal = e.Value
		return true

	case *parser.BinaryExpr:
		return evalBinaryExpr(e, result)

	case *parser.UnaryExpr:
		return evalUnaryExpr(e, result)

	default:
		return false
	}
}

// evalBinaryExpr evaluates a binary expression
func evalBinaryExpr(expr *parser.BinaryExpr, result *ConstValue) bool {
	var leftVal, rightVal ConstValue

	leftConst := EvalExpr(expr.Left, &leftVal)
	rightConst := EvalExpr(expr.Right, &rightVal)

	if !leftConst || !rightConst {
		return false
	}

	op := lexer.TokenType(expr.Op)

	// Integer operations
	if leftVal.Type == ConstValInt && rightVal.Type == ConstValInt {
		a := leftVal.Value.IntVal
		b := rightVal.Value.IntVal

		switch op {
		case lexer.TOKEN_PLUS:
			result.Type = ConstValInt
			result.Value.IntVal = a + b
			return true

		case lexer.TOKEN_MINUS:
			result.Type = ConstValInt
			result.Value.IntVal = a - b
			return true

		case lexer.TOKEN_ASTERISK:
			result.Type = ConstValInt
			result.Value.IntVal = a * b
			return true

		case lexer.TOKEN_SLASH:
			if b == 0 {
				return false // Division by zero
			}
			result.Type = ConstValInt
			result.Value.IntVal = a / b
			return true

		case lexer.TOKEN_PERCENT:
			if b == 0 {
				return false // Division by zero
			}
			result.Type = ConstValInt
			result.Value.IntVal = a % b
			return true

		case lexer.TOKEN_EQUAL:
			result.Type = ConstValBool
			result.Value.BoolVal = (a == b)
			return true

		case lexer.TOKEN_NOT_EQUAL:
			result.Type = ConstValBool
			result.Value.BoolVal = (a != b)
			return true

		case lexer.TOKEN_LESS:
			result.Type = ConstValBool
			result.Value.BoolVal = (a < b)
			return true

		case lexer.TOKEN_LESS_EQUAL:
			result.Type = ConstValBool
			result.Value.BoolVal = (a <= b)
			return true

		case lexer.TOKEN_GREATER:
			result.Type = ConstValBool
			result.Value.BoolVal = (a > b)
			return true

		case lexer.TOKEN_GREATER_EQUAL:
			result.Type = ConstValBool
			result.Value.BoolVal = (a >= b)
			return true

		default:
			return false
		}
	}

	// Float operations (convert int to float if needed)
	if leftVal.Type == ConstValFloat || rightVal.Type == ConstValFloat {
		var a, b float64
		if leftVal.Type == ConstValFloat {
			a = leftVal.Value.FloatVal
		} else {
			a = float64(leftVal.Value.IntVal)
		}
		if rightVal.Type == ConstValFloat {
			b = rightVal.Value.FloatVal
		} else {
			b = float64(rightVal.Value.IntVal)
		}

		switch op {
		case lexer.TOKEN_PLUS:
			result.Type = ConstValFloat
			result.Value.FloatVal = a + b
			return true

		case lexer.TOKEN_MINUS:
			result.Type = ConstValFloat
			result.Value.FloatVal = a - b
			return true

		case lexer.TOKEN_ASTERISK:
			result.Type = ConstValFloat
			result.Value.FloatVal = a * b
			return true

		case lexer.TOKEN_SLASH:
			if b == 0.0 {
				return false // Division by zero
			}
			result.Type = ConstValFloat
			result.Value.FloatVal = a / b
			return true

		case lexer.TOKEN_EQUAL:
			result.Type = ConstValBool
			result.Value.BoolVal = (a == b)
			return true

		case lexer.TOKEN_NOT_EQUAL:
			result.Type = ConstValBool
			result.Value.BoolVal = (a != b)
			return true

		case lexer.TOKEN_LESS:
			result.Type = ConstValBool
			result.Value.BoolVal = (a < b)
			return true

		case lexer.TOKEN_LESS_EQUAL:
			result.Type = ConstValBool
			result.Value.BoolVal = (a <= b)
			return true

		case lexer.TOKEN_GREATER:
			result.Type = ConstValBool
			result.Value.BoolVal = (a > b)
			return true

		case lexer.TOKEN_GREATER_EQUAL:
			result.Type = ConstValBool
			result.Value.BoolVal = (a >= b)
			return true

		default:
			return false
		}
	}

	return false
}

// evalUnaryExpr evaluates a unary expression
func evalUnaryExpr(expr *parser.UnaryExpr, result *ConstValue) bool {
	var operandVal ConstValue

	if !EvalExpr(expr.Operand, &operandVal) {
		return false
	}

	op := lexer.TokenType(expr.Op)

	switch op {
	case lexer.TOKEN_MINUS:
		if operandVal.Type == ConstValInt {
			result.Type = ConstValInt
			result.Value.IntVal = -operandVal.Value.IntVal
			return true
		} else if operandVal.Type == ConstValFloat {
			result.Type = ConstValFloat
			result.Value.FloatVal = -operandVal.Value.FloatVal
			return true
		}
		return false

	case lexer.TOKEN_EXCLAMATION:
		if operandVal.Type == ConstValBool {
			result.Type = ConstValBool
			result.Value.BoolVal = !operandVal.Value.BoolVal
			return true
		}
		return false

	default:
		return false
	}
}

// IsConstant checks if an expression is a constant
func IsConstant(expr parser.Expr) bool {
	if expr == nil {
		return false
	}

	switch expr.(type) {
	case *parser.NumberLiteral, *parser.BoolLiteral:
		return true

	case *parser.BinaryExpr:
		e := expr.(*parser.BinaryExpr)
		return IsConstant(e.Left) && IsConstant(e.Right)

	case *parser.UnaryExpr:
		e := expr.(*parser.UnaryExpr)
		return IsConstant(e.Operand)

	default:
		return false
	}
}

// GetIntValue gets the integer value of a constant expression
// Returns 0 if the expression is not a constant or not an integer
func GetIntValue(expr parser.Expr) int64 {
	if expr == nil {
		return 0
	}

	switch e := expr.(type) {
	case *parser.NumberLiteral:
		val, err := strconv.ParseInt(e.Value, 10, 64)
		if err == nil {
			return val
		}
		return 0

	case *parser.BoolLiteral:
		if e.Value {
			return 1
		}
		return 0

	default:
		var result ConstValue
		if EvalExpr(expr, &result) && result.Type == ConstValInt {
			return result.Value.IntVal
		}
		return 0
	}
}

